<a href="https://colab.research.google.com/github/LumaLynai/ProjetoFinalETL/blob/main/ProjetoFinalETL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Bibliotecas**

In [ ]:
!pip install pandas
!pip install gcsfs


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 140 kB 3.6 MB/s 
     |████████████████████████████████| 1.1 MB 43.5 MB/s 
     |████████████████████████████████| 271 kB 45.2 MB/s 
     |████████████████████████████████| 144 kB 65.4 MB/s 
     |████████████████████████████████| 94 kB 3.1 MB/s 


In [ ]:
!pip install plotly

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
pd.options.plotting.backend='plotly'

In [ ]:
!pip install pyspark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 281.3 MB 43 kB/s 
     |████████████████████████████████| 199 kB 64.6 MB/s 
  Created wheel for pyspark: filename=pyspark-3.3.0-py2.py3-none-any.whl size=281764026 sha256=1bea4a3f3ca4a54f3df37f617aa996133df83f8d50ad8c28c7fc1f1331d96958
  Stored in directory: /root/.cache/pip/wheels/7a/8e/1b/f73a52650d2e5f337708d9f6a1750d451a7349a867f928b885
Successfully built pyspark


In [ ]:
import pandas as pd
import psycopg2
import requests
import pymongo
import gcsfs
import os 
import csv

/usr/local/lib/python3.7/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


In [ ]:
from matplotlib import pyplot as plt

In [ ]:
from google.cloud import storage

In [ ]:
from pyspark.sql import SparkSession
from pyspark.sql.window import *
from pyspark.sql.types import *
from pyspark.sql.group import *
from pyspark import SparkConf
import pyspark.sql.functions as F 
import pyspark.sql.window as w

# **GCP**
Conectando backet

In [ ]:
serviceAccount = r'/content/drive/MyDrive/arquivos/macro-mercury-349020-d9ed9a670580 (1).json'
os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = serviceAccount

In [ ]:
client = storage.Client()
bucket = client.get_bucket('soulcode-bc17-telecom')

#**Meu_Municio_Acesso.csv**

```
```



##**Pandas**


In [ ]:
df1_extrai = pd.read_csv('https://storage.googleapis.com/telecomunicacao2/Meu_Municipio_Acessos.csv',sep=';')

**Copy** 

In [ ]:
df1_pd = df1_extrai.copy()

### Inserindo no Backet 
Original

In [ ]:
from pandas._libs import index
df1_extrai.to_csv('gs://soulcode-bc17-telecom/Original/Meu_Municipio_Acessos.csv',index=False)


In [ ]:
df1_pd.head(2)

### Info

In [ ]:
df1_pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66837 entries, 0 to 66836
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Ano              66837 non-null  int64 
 1   Mês              66837 non-null  int64 
 2   Acessos          66837 non-null  int64 
 3   Serviço          66837 non-null  object
 4   Densidade        66837 non-null  object
 5   Código IBGE      66837 non-null  int64 
 6   Município        66837 non-null  object
 7   UF               66837 non-null  object
 8   Nome UF          66837 non-null  object
 9   Região           66837 non-null  object
 10  Código Nacional  66837 non-null  int64 
dtypes: int64(5), object(6)
memory usage: 5.6+ MB


### Unique

In [ ]:
df1_pd['Município'].unique()

array(['Paulo Frontin - PR', 'Araruama - RJ', 'Pinhão - PR', ...,
       'Ribeiro Gonçalves - PI', 'Itacurubi - RS', 'Carrapateira - PB'],
      dtype=object)

In [ ]:
df1=df1_pd['Município'].str[:-5]

In [ ]:
df1_pd['Município']=df1_pd['Município'].str[:-5]

In [ ]:
df1_pd.head()

Ano  Mês  Acessos            Serviço        Densidade  Código IBGE  \
0  2021    6     5597    Telefonia Móvel  73,114931636659      4118709   
1  2021    6      713   Banda Larga Fixa  27,993718099725      4118709   
2  2021    6      132  TV por Assinatura  5,1825677267373      4118709   
3  2021    6      327     Telefonia Fixa   12,83863368669      4118709   
4  2021    6   162149    Telefonia Móvel  113,77138048893      3300209   

       Município  UF         Nome UF   Região  Código Nacional  
0  Paulo Frontin  PR          Paraná      Sul               42  
1  Paulo Frontin  PR          Paraná      Sul               42  
2  Paulo Frontin  PR          Paraná      Sul               42  
3  Paulo Frontin  PR          Paraná      Sul               42  
4       Araruama  RJ  Rio de Janeiro  Sudeste               22

##**Spark**

###SessionSpark


In [ ]:
spark = SparkSession.builder.appName("pandas em spark").getOrCreate()

In [ ]:
spark

In [ ]:
df1_spark = spark.createDataFrame(df1_pd)

In [ ]:
df1_spark.show()

+----+---+-------+-----------------+----------------+-----------+--------------------+---+--------------+--------+---------------+
| Ano|Mês|Acessos|          Serviço|       Densidade|Código IBGE|           Município| UF|       Nome UF|  Região|Código Nacional|
+----+---+-------+-----------------+----------------+-----------+--------------------+---+--------------+--------+---------------+
|2021|  6|   5597|  Telefonia Móvel| 73,114931636659|    4118709|  Paulo Frontin - PR| PR|        Paraná|     Sul|             42|
|2021|  6|    713| Banda Larga Fixa| 27,993718099725|    4118709|  Paulo Frontin - PR| PR|        Paraná|     Sul|             42|
|2021|  6|    132|TV por Assinatura| 5,1825677267373|    4118709|  Paulo Frontin - PR| PR|        Paraná|     Sul|             42|
|2021|  6|    327|   Telefonia Fixa|  12,83863368669|    4118709|  Paulo Frontin - PR| PR|        Paraná|     Sul|             42|
|2021|  6| 162149|  Telefonia Móvel| 113,77138048893|    3300209|       Araruama - 

### Distinct

In [ ]:
df1_spark.select('Código Nacional').distinct().count()

67

In [ ]:
df1_spark.select('Código Nacional').distinct().show()

+---------------+
|Código Nacional|
+---------------+
|             65|
|             19|
|             54|
|             22|
|             77|
|             34|
|             94|
|             43|
|             32|
|             84|
|             31|
|             98|
|             95|
|             71|
|             68|
|             87|
|             63|
|             27|
|             51|
|             17|
+---------------+
only showing top 20 rows



In [ ]:
df1_spark.select('Município').distinct().show()

+--------------------+
|           Município|
+--------------------+
|            Araruama|
|        Borrazópolis|
|              Utinga|
|              Maruim|
|             Ibiporã|
|          Livramento|
|    Senador Guiomard|
|           Fronteira|
|              Pongaí|
| Leopoldo de Bulhões|
|               Apodi|
|      Benedito Leite|
|   Tiradentes do Sul|
|        Assis Brasil|
|  Ribas do Rio Pardo|
|Santo Antônio dos...|
|Boa Esperança do Sul|
|           Carrancas|
|         Piranguinho|
|            Rio Novo|
+--------------------+
only showing top 20 rows



### Like

In [ ]:
df1_spark.filter(F.col('Município').like("%Santo Antônio dos%")).show()

+----+---+-------+-----------------+---------------+-----------+--------------------+---+--------+--------+
| Ano|Mês|Acessos|          Serviço|      Densidade|Código IBGE|           Município| UF| Nome UF|  Região|
+----+---+-------+-----------------+---------------+-----------+--------------------+---+--------+--------+
|2021|  6|  11480|  Telefonia Móvel|77,558187577469|    2110302|Santo Antônio dos...| MA|Maranhão|Nordeste|
|2021|  6|     54| Banda Larga Fixa|1,3015184381779|    2110302|Santo Antônio dos...| MA|Maranhão|Nordeste|
|2021|  6|    428|TV por Assinatura|10,315738732225|    2110302|Santo Antônio dos...| MA|Maranhão|Nordeste|
|2021|  6|     46|   Telefonia Fixa|1,1087008917812|    2110302|Santo Antônio dos...| MA|Maranhão|Nordeste|
|2021|  6|   1914|  Telefonia Móvel|87,165281625115|    2209450|Santo Antônio dos...| PI|   Piauí|Nordeste|
|2021|  6|     93| Banda Larga Fixa|14,176829268293|    2209450|Santo Antônio dos...| PI|   Piauí|Nordeste|
|2021|  6|     82|TV por Ass

In [ ]:
df1_spark.filter(F.col('Município').like("%Couto de Magalhãe%")).show()

+----+---+-------+-----------------+---------------+-----------+--------------------+---+------------+-------+
| Ano|Mês|Acessos|          Serviço|      Densidade|Código IBGE|           Município| UF|     Nome UF| Região|
+----+---+-------+-----------------+---------------+-----------+--------------------+---+------------+-------+
|2021|  6|   3390|  Telefonia Móvel|73,841284196247|    3120102|Couto de Magalhãe...| MG|Minas Gerais|Sudeste|
|2021|  6|    372| Banda Larga Fixa|25,237449118046|    3120102|Couto de Magalhãe...| MG|Minas Gerais|Sudeste|
|2021|  6|    177|TV por Assinatura|12,008141112619|    3120102|Couto de Magalhãe...| MG|Minas Gerais|Sudeste|
|2021|  6|     91|   Telefonia Fixa|6,1736770691995|    3120102|Couto de Magalhãe...| MG|Minas Gerais|Sudeste|
|2020| 12|   2598|  Telefonia Móvel|58,738412841962|    3120102|Couto de Magalhãe...| MG|Minas Gerais|Sudeste|
|2020| 12|     69| Banda Larga Fixa|4,6811397557666|    3120102|Couto de Magalhãe...| MG|Minas Gerais|Sudeste|
|

In [ ]:
df1_spark.filter(F.col('Município').like("%MG%")).count()

10235

### Drop coluna
* Código Nacional
* Nome UF

In [ ]:
df1_spark=df1_spark.drop(F.col('Código Nacional')).drop(F.col('Nome UF'))

In [ ]:
df1_spark.show(1)

+----+---+-------+---------------+---------------+-----------+------------------+---+------+
| Ano|Mês|Acessos|        Serviço|      Densidade|Código IBGE|         Município| UF|Região|
+----+---+-------+---------------+---------------+-----------+------------------+---+------+
|2021|  6|   5597|Telefonia Móvel|73,114931636659|    4118709|Paulo Frontin - PR| PR|   Sul|
+----+---+-------+---------------+---------------+-----------+------------------+---+------+
only showing top 1 row



###witcolumnRenamed
* Mês == Mes
* Serviço == Servico
* Código IBGE == Cod_IBGE
* Município == Municipio
* Região == Regiao

In [ ]:
df1_spark=(df1_spark.withColumnRenamed("Mês","Mes")
         .withColumnRenamed("Serviço","Servico")
         .withColumnRenamed("Código IBGE","Cod_IBGE")
         .withColumnRenamed("Município","Municipio")
         .withColumnRenamed("Região","Regiao")
)

In [ ]:
df1_spark.show()

+----+---+-------+-----------------+----------------+--------+--------------------+---+--------------+--------+
| Ano|Mes|Acessos|          Servico|       Densidade|Cod_IBGE|           Municipio| UF|       Nome UF|  Regiao|
+----+---+-------+-----------------+----------------+--------+--------------------+---+--------------+--------+
|2021|  6|   5597|  Telefonia Móvel| 73,114931636659| 4118709|  Paulo Frontin - PR| PR|        Paraná|     Sul|
|2021|  6|    713| Banda Larga Fixa| 27,993718099725| 4118709|  Paulo Frontin - PR| PR|        Paraná|     Sul|
|2021|  6|    132|TV por Assinatura| 5,1825677267373| 4118709|  Paulo Frontin - PR| PR|        Paraná|     Sul|
|2021|  6|    327|   Telefonia Fixa|  12,83863368669| 4118709|  Paulo Frontin - PR| PR|        Paraná|     Sul|
|2021|  6| 162149|  Telefonia Móvel| 113,77138048893| 3300209|       Araruama - RJ| RJ|Rio de Janeiro| Sudeste|
|2021|  6|  32457| Banda Larga Fixa| 67,673734909614| 3300209|       Araruama - RJ| RJ|Rio de Janeiro| S

###Inserindo bucket 
Tratados 

In [ ]:
df1_spark.toPandas().to_parquet('gs://soulcode-bc17-telecom/Tratados/Meu_Municipio_AcessosTratad.gzip',compression='gzip')

# **Velocidade_Contratada_SCM.csv**


In [ ]:
df2_extrai=pd.read_csv('gs://telecomunicacao2/Velocidade_Contratada_SCM.csv',sep=";")

**Copy**

In [ ]:
df2_pd = df2_extrai.copy()

###Inserindo Bucket

In [ ]:
df2_extrai.to_csv('gs://soulcode-bc17-telecom/Original/Velocidade_Contratada_SCM.csv',index=False)


In [ ]:
df2_pd.head(2)

Ano   Mês       razao_social                cnpj  \
0  2020.0  12.0  algar telecom s/a  71.208.516/0001-74   
1  2020.0  12.0  algar telecom s/a  71.208.516/0001-74   

  velocidade_contratada_mbps  uf   municipio  codigo_ibge  acessos      tipo  \
0                         40  MG  UBERLANDIA    3170206.0  16660.0  INTERNET   
1                         40  SP      FRANCA    3516200.0   5093.0  INTERNET   

      municipio_uf  
0  Uberlândia - MG  
1      Franca - SP

###Unique

In [ ]:
pd.unique(df2_pd['Mês'])

array([12., 11., 10.,  9.,  8.,  7.,  6.,  5.,  4.,  3.,  2.,  1., nan])

###Info

In [ ]:
df2_pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5789372 entries, 0 to 5789371
Data columns (total 11 columns):
 #   Column                      Dtype  
---  ------                      -----  
 0   Ano                         float64
 1   Mês                         float64
 2   razao_social                object 
 3   cnpj                        object 
 4   velocidade_contratada_mbps  object 
 5   uf                          object 
 6   municipio                   object 
 7   codigo_ibge                 float64
 8   acessos                     float64
 9   tipo                        object 
 10  municipio_uf                object 
dtypes: float64(4), object(7)
memory usage: 485.9+ MB


###Duplicated

In [ ]:
df2_pd.duplicated().sum()

11684

In [ ]:
df_pd=df2_pd[df2_pd.duplicated()]

In [ ]:
df_pd.head()

In [ ]:
df2_pd= df2_pd.drop_duplicates()

In [ ]:
df2_pd.count()

Ano                           5777687
Mês                           5777687
razao_social                  5777687
cnpj                          5777687
velocidade_contratada_mbps    5777593
uf                            5777687
municipio                     5777687
codigo_ibge                   5777687
acessos                       5777432
tipo                          5777687
municipio_uf                  5777687
dtype: int64

###Isnull

In [ ]:
pd.isnull(df2_pd).sum()

Ano                             1
Mês                             1
razao_social                    1
cnpj                            1
velocidade_contratada_mbps     95
uf                              1
municipio                       1
codigo_ibge                     1
acessos                       256
tipo                            1
municipio_uf                    1
dtype: int64

###Drop

**Coluna**

In [ ]:
df2_pd=df2_pd.drop(columns=['municipio_uf'])

In [ ]:
df2_pd.head()

Ano   Mes       razao_social                cnpj  \
0  2020.0  12.0  algar telecom s/a  71.208.516/0001-74   
1  2020.0  12.0  algar telecom s/a  71.208.516/0001-74   
2  2020.0  12.0  algar telecom s/a  71.208.516/0001-74   
3  2020.0  12.0  algar telecom s/a  71.208.516/0001-74   
4  2020.0  12.0  algar telecom s/a  71.208.516/0001-74   

  velocidade_contratada_mbps  uf      municipio   Cod_IBGE  acessos      tipo  
0                         40  MG     UBERLANDIA  3170206.0  16660.0  INTERNET  
1                         40  SP         FRANCA  3516200.0   5093.0  INTERNET  
2                         10  MG     UBERLANDIA  3170206.0  10507.0  INTERNET  
3                         60  MG    CAPINOPOLIS  3112604.0    498.0  INTERNET  
4                        100  MG  SANTA JULIANA  3157708.0    508.0  INTERNET

**Linhas** 
* Valores  isnull == 360
* df2_pd == 5777687
* isnull 0.006% == df2_pd

In [ ]:
df2_pd=df2_pd.dropna()

In [ ]:
pd.isnull(df2_pd).sum()

Ano                           0
Mês                           0
razao_social                  0
cnpj                          0
velocidade_contratada_mbps    0
uf                            0
municipio                     0
codigo_ibge                   0
acessos                       0
tipo                          0
municipio_uf                  0
dtype: int64

###Unique

In [ ]:
pd.unique(df2_pd['municipio_uf'])

array(['Uberlândia - MG', 'Franca - SP', 'Capinópolis - MG', ...,
       'Quixaba - PB', 'Quixaba - PE', 'São Valério - TO'], dtype=object)

###Rename 


In [ ]:
df2_pd.rename(columns = {'Mês':'Mes','codigo_ibge':'Cod_IBGE'},inplace=True)

In [ ]:
df2_pd.head(1)

Ano   Mes       razao_social                cnpj  \
0  2020.0  12.0  algar telecom s/a  71.208.516/0001-74   

  velocidade_contratada_mbps  uf   municipio   Cod_IBGE  acessos      tipo  
0                         40  MG  UBERLANDIA  3170206.0  16660.0  INTERNET

###Inserindo bucket 
Tratados

In [ ]:
df2_pd.to_parquet('gs://soulcode-bc17-telecom/Tratados/Velocidade_Contratada_Tratad.gzip',compression='gzip')

In [ ]:
df2_pd.head(1)

Ano   Mês       razao_social                cnpj  \
0  2020.0  12.0  algar telecom s/a  71.208.516/0001-74   

  velocidade_contratada_mbps  uf   municipio  codigo_ibge  acessos      tipo  \
0                         40  MG  UBERLANDIA    3170206.0  16660.0  INTERNET   

      municipio_uf  
0  Uberlândia - MG

# **Acessos_Telefonia_Movel_Pre_Pos_Total.json**

In [ ]:
df3_extrai = pd.read_json('gs://soulcode-bc17-telecom/Acessos_Telefonia_Movel_Pre_Pos_Total.json')

In [ ]:
df3_pd=df3_extrai.copy()

In [ ]:
df3_pd.head()

Ano  Mês Modalidade de Cobrança    Acessos
0  2022    3               Pós-pago  138629611
1  2022    3               Pré-pago  119642952
2  2022    2               Pós-pago  137439913
3  2022    2               Pré-pago  119004007
4  2022    1               Pós-pago  136500794

## Duplicated

In [ ]:
df3_test=df3_pd[df3_pd.duplicated()]

In [ ]:
df3_test.count()

Ano                       0
Mês                       0
Modalidade de Cobrança    0
Acessos                   0
dtype: int64

## Info

In [ ]:
df3_pd.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 412 entries, 0 to 411
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Ano                     412 non-null    int64 
 1   Mês                     412 non-null    int64 
 2   Modalidade de Cobrança  412 non-null    object
 3   Acessos                 412 non-null    int64 
dtypes: int64(3), object(1)
memory usage: 16.1+ KB


## Unique 

In [ ]:
pd.unique(df3_pd['Mês'])

array([ 3,  2,  1, 12, 11, 10,  9,  8,  7,  6,  5,  4])

## Rename

In [ ]:
df3_pd.rename(columns = {'Mês':'Mes','Modalidade de Cobrança':'ModalidadeCobranca'},inplace=True)

In [ ]:
df3_pd.head()

## Inserindo Bucket

In [ ]:
df3_pd.to_parquet('gs://soulcode-bc17-telecom/Tratados/Acessos_Telefonia_Movel_Pre_Pos_TotalTratad.gzip',compression='gzip')

#**Densidade_Telefonia_Movel.json**

In [ ]:
df4_extrai = pd.read_json('gs://soulcode-bc17-telecom/Densidade_Telefonia_Movel.json')

In [ ]:
df4_pd = df4_extrai.copy()

In [ ]:
df4_pd.count()

Ano                           223046
Mês                           223046
UF                            222998
Município                     222998
Código IBGE                   222998
Densidade                     222998
Nível Geográfico Densidade    223046
dtype: int64

In [ ]:
df4_pd.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 223046 entries, 0 to 223045
Data columns (total 7 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Ano                         223046 non-null  int64  
 1   Mês                         223046 non-null  int64  
 2   UF                          222998 non-null  object 
 3   Município                   222998 non-null  object 
 4   Código IBGE                 222998 non-null  float64
 5   Densidade                   222998 non-null  object 
 6   Nível Geográfico Densidade  223046 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 13.6+ MB


In [ ]:
pd.unique(df4_pd['Nível Geográfico Densidade'])

array(['Brasil', 'UF', 'Municipio'], dtype=object)

In [ ]:
df4_test=df4_pd[df4_pd.duplicated()]

In [ ]:
df4_test.count()

Ano                           0
Mês                           0
UF                            0
Município                     0
Código IBGE                   0
Densidade                     0
Nível Geográfico Densidade    0
dtype: int64

## Rename

In [ ]:
df4_pd.rename(columns = {'Mês':'Mes','Município':'Municipio','Código IBGE':'Cod_IBGE','Nível Geográfico Densidade':'NivelGeografico'},inplace=True)

In [ ]:
df4_pd.head()

## Inserindo no Bucket 

In [ ]:
df4_pd.to_parquet('gs://soulcode-bc17-telecom/Tratados/Densidade_Telefonia_Movel.gzip',compression='gzip')

#**Select**
Spark SQL

In [ ]:
spark = (
    SparkSession.builder
                .master('local')
                .appName('spark-gcs')
                .config('spark.ui.port', '4050')
                .config("spark.jars", 'https://storage.googleapis.com/hadoop-lib/gcs/gcs-connector-hadoop2-latest.jar')
                .getOrCreate()
)

### Acessos_Banda_Larga_FixaTratad.parquet
df1_spark

In [ ]:
df1_sp = (spark.read.format("parquet")
                .option("header", True)
                .option("delimiter", ",")
                .load("gs://soulcode-bc17-telecom/Tratados/Acessos_Banda_Larga_FixaTratad.parquet")
)

In [ ]:
df1_sp.select('Ano').distinct().show()

+----+
| Ano|
+----+
|2022|
+----+



In [ ]:
df1_sp.select('empresa').distinct().show()

In [ ]:
df1_sp.createOrReplaceTempView("bandfixa")

In [ ]:
spark.sql("select * from bandfixa").show(2)

+----+---+--------------+--------------------+--------------+---------------+---+-------------+--------+---------------+----------+-------------+---------------+-------+-----------------+
| Ano|Mes|GrupoEconomico|             Empresa|          CNPJ|PortePrestadora| UF|    Municipio|Cod_IBGE|FaixaVelocidade|Tecnologia|   MeioAcesso|     TipoPessoa|Acessos|__index_level_0__|
+----+---+--------------+--------------------+--------------+---------------+---+-------------+--------+---------------+----------+-------------+---------------+-------+-----------------+
|2022|  3|            OI|                  OI|76535764000143|   Grande Porte| RS|Caxias do Sul| 4305108|512kbps a 2Mbps|     ADSL1|Cabo Metálico|Pessoa Jurídica|      2|                0|
|2022|  3|        OUTROS|Telecomunicacoes ...|  336701000104|  Pequeno Porte| MG|    Rio Acima| 3154804|512kbps a 2Mbps|      VSAT|     Satélite|Pessoa Jurídica|      1|                1|
+----+---+--------------+--------------------+--------------

#### Quantos acessos por UF decrescente 

In [ ]:
spark.sql("select uf,sum(acessos) as acessos From bandfixa group by uf order by acessos desc").show()

+---+--------+
| uf| acessos|
+---+--------+
| SP|38570157|
| MG|13497977|
| RJ|11003963|
| RS| 8741519|
| PR| 8683036|
| SC| 6876966|
| BA| 4502075|
| CE| 3942285|
| GO| 3527592|
| PE| 2493138|
| DF| 2388766|
| ES| 2240146|
| MT| 1949371|
| PA| 1864320|
| RN| 1803992|
| PB| 1672301|
| MS| 1467560|
| MA| 1379286|
| AM| 1352689|
| PI| 1067413|
+---+--------+
only showing top 20 rows



#####Pandas graficos 

In [ ]:
df1pd = spark.sql("select uf,sum(acessos) as acessos From bandfixa group by uf order by acessos desc")

In [ ]:
df1pd=df1pd.toPandas()

In [ ]:
df1pd.head(1)

uf   acessos
0  SP  38570157

In [ ]:
df1pd.plot(x='uf', y='acessos',kind='barh',figsize=(15,7),sharey=True)

#### Porcentagem de acessos por UF

In [ ]:
df1_sp.createOrReplaceTempView("bandfixa")

In [ ]:
spark.sql("select uf,sum(acessos) as acessos,round((sum(acessos)/122858439)*100,2) as porcent From bandfixa group by uf order by porcent desc ").show(3)

+---+--------+-------+
| uf| acessos|porcent|
+---+--------+-------+
| SP|38570157|  31.39|
| MG|13497977|  10.99|
| RJ|11003963|   8.96|
+---+--------+-------+
only showing top 3 rows



#####Pandas graficos 

In [ ]:
df12pd=spark.sql("select uf,sum(acessos) as acessos,round((sum(acessos)/122858439)*100,2) as porcent From bandfixa group by uf order by porcent desc ")

In [ ]:
df12pd=df12pd.toPandas()

In [ ]:
df12pd.plot(x='uf', y='porcent',kind='bar',color='porcent')

###Porcentagem de acessos por empresa

In [ ]:
spark.sql("select empresa,sum(acessos) as acessos,round((sum(acessos)/122858439)*100,2) as porcent From bandfixa group by empresa order by porcent desc ").show(5)

+--------------------+--------+-------+
|             empresa| acessos|porcent|
+--------------------+--------+-------+
|               CLARO|29138258|  23.72|
|                VIVO|18949934|  15.42|
|                  OI|15426785|  12.56|
|Brisanet Servicos...| 2654973|   2.16|
|ALGAR (CTBC TELECOM)| 2253743|   1.83|
+--------------------+--------+-------+
only showing top 5 rows



#####Pandas graficos 


In [ ]:
spark.sql("select empresa,sum(acessos) as acessos,round((sum(acessos)/122858439)*100,2) as porcent From bandfixa where acessos between 29138258 and 15426785   group by empresa order by porcent desc ").show()

+-------+-------+-------+
|empresa|acessos|porcent|
+-------+-------+-------+
+-------+-------+-------+



In [ ]:
df13pd=df13pd.toPandas()

In [ ]:
df13=df13pd['acessos']>=2253743

In [ ]:
df13.head()

0    True
1    True
2    True
3    True
4    True
Name: acessos, dtype: bool

In [ ]:
df13.plot(x='empresa',y='porcent',kind='bar')

### Porcentagem de acessos por tecnologia e meio de acesso 

In [ ]:
spark.sql("select Tecnologia,meioacesso, sum(acessos) as acessos,round((sum(acessos)/122858439)*100,2) as porcent From bandfixa group by tecnologia, meioacesso  order by porcent desc ").show(7)

+----------+-------------+--------+-------+
|Tecnologia|   meioacesso| acessos|porcent|
+----------+-------------+--------+-------+
|      FTTH|        Fibra|72030841|  58.63|
|       HFC| Cabo Coaxial|27106018|  22.06|
|     ADSL2|Cabo Metálico| 7497119|    6.1|
|  ETHERNET|        Fibra| 4056768|    3.3|
|     Wi-Fi|        Rádio| 1816009|   1.48|
|  ETHERNET|        Rádio| 1704919|   1.39|
|  ETHERNET|Cabo Metálico| 1541481|   1.25|
+----------+-------------+--------+-------+
only showing top 7 rows



### Porcentagem acessos por meio de acessos 

In [ ]:
spark.sql("select meioacesso, sum(acessos) as acessos,round((sum(acessos)/122858439)*100,2) as porcent From bandfixa group by meioacesso  order by porcent desc ").show(7)

+-------------+--------+-------+
|   meioacesso| acessos|porcent|
+-------------+--------+-------+
|        Fibra|78092548|  63.56|
| Cabo Coaxial|27337039|  22.25|
|Cabo Metálico|11470946|   9.34|
|        Rádio| 5043698|   4.11|
|     Satélite|  914208|   0.74|
+-------------+--------+-------+



### Porcentagem de acessos por faixa de velocidade 

In [ ]:
spark.sql("select faixavelocidadde ,sum(acessos) as acessos,round((sum(acessos)/122858439)*100,2) as porcent From bandfixa group by faixavelocidade order by porcent desc ").show(7)

**Porcentagem de acessos por UF a velocidade > 34Mbps**

In [ ]:
spark.sql("select uf,sum(acessos) as acessos,round((sum(acessos)/99320322)*100,2) as porcent From bandfixa where faixavelocidade='> 34Mbps' group by uf order by porcent desc ").show(5)

+---+--------+-------+
| uf| acessos|porcent|
+---+--------+-------+
| SP|32548757|  32.77|
| MG|10715173|  10.79|
| RJ| 9181128|   9.24|
| RS| 6959905|   7.01|
| PR| 6847625|   6.89|
+---+--------+-------+
only showing top 5 rows



**Porcentagem de acessos por UF a velocidade 12Mbps a 34Mbps**




In [ ]:
spark.sql("select uf,sum(acessos) as acessos,round((sum(acessos)/9857714)*100,2) as porcent From bandfixa where faixavelocidade='12Mbps a 34Mbps' group by uf order by porcent desc ").show(5)

+---+-------+-------+
| uf|acessos|porcent|
+---+-------+-------+
| SP|2014711|  20.44|
| MG|1144589|  11.61|
| BA| 884924|   8.98|
| PR| 851667|   8.64|
| RJ| 850718|   8.63|
+---+-------+-------+
only showing top 5 rows



**Porcentagem de acessos por UF a velocidade 2Mbps a 12Mbps**


In [ ]:
spark.sql("select uf,sum(acessos) as acessos,round((sum(acessos)/9510361)*100,2) as porcent From bandfixa where faixavelocidade='2Mbps a 12Mbps' group by uf order by porcent desc ").show(5)

+---+-------+-------+
| uf|acessos|porcent|
+---+-------+-------+
| SP|2274833|  23.92|
| MG|1265958|  13.31|
| RS| 815782|   8.58|
| PR| 729387|   7.67|
| RJ| 644927|   6.78|
+---+-------+-------+
only showing top 5 rows



**Porcentagem de acessos por UF a velocidade 512kbps a 2Mbps**




In [ ]:
spark.sql("select uf,sum(acessos) as acessos,round((sum(acessos)/3859448)*100,2) as porcent From bandfixa where faixavelocidade='512kbps a 2Mbps' group by uf order by porcent desc ").show(5)

+---+-------+-------+
| uf|acessos|porcent|
+---+-------+-------+
| SP|1684775|  43.65|
| MG| 337801|   8.75|
| RJ| 291216|   7.55|
| PR| 233830|   6.06|
| RS| 197523|   5.12|
+---+-------+-------+
only showing top 5 rows



**Porcentagem de acessos por UF a velocidade 0Kbps a 512Kbps**

In [ ]:
spark.sql("select uf,sum(acessos) as acessos,round((sum(acessos)/310594)*100,2) as porcent From bandfixa where faixavelocidade='0Kbps a 512Kbps' group by uf order by porcent desc ").show(5)

+---+-------+-------+
| uf|acessos|porcent|
+---+-------+-------+
| SP|  47081|  15.16|
| RJ|  35974|  11.58|
| MG|  34456|  11.09|
| RS|  28143|   9.06|
| PR|  20527|   6.61|
+---+-------+-------+
only showing top 5 rows



###Porcentagem de acesso em SP a faixa de velocidade

In [ ]:
#@title Texto de título padrão
spark.sql("select  faixavelocidade, sum(acessos) as acessos,round((sum(acessos)/38570157)*100,2) as porcent From bandfixa where uf='SP' group by faixavelocidade order by porcent desc ").show(7)

+---------------+--------+-------+
|faixavelocidade| acessos|porcent|
+---------------+--------+-------+
|       > 34Mbps|32548757|  84.39|
| 2Mbps a 12Mbps| 2274833|    5.9|
|12Mbps a 34Mbps| 2014711|   5.22|
|512kbps a 2Mbps| 1684775|   4.37|
|0Kbps a 512Kbps|   47081|   0.12|
+---------------+--------+-------+



Ordenando Faixa de Velocidade no estado de SP no mês 3 

In [ ]:
spark.sql("select Ano, Mes,Empresa , UF ,FaixaVelocidade from bandfixa where Mes == '3' and Uf = 'SP' order by FaixaVelocidade ;").show(10)

contagem de tecnologia por uf

In [ ]:
spark.sql("select UF, tecnologia, count(Tecnologia) as countTec from bandfixa group by uf, tecnologia order by countTec desc ;").show()

+---+----------+--------+
| UF|tecnologia|countTec|
+---+----------+--------+
| SP|      FTTH|   93092|
| MG|      FTTH|   76254|
| RS|      FTTH|   51832|
| SP|  ETHERNET|   44379|
| SC|      FTTH|   44051|
| RS|  ETHERNET|   38227|
| PR|      FTTH|   37775|
| SP|     ADSL2|   29972|
| MG|  ETHERNET|   23393|
| MG|      VSAT|   20858|
| PR|  ETHERNET|   20609|
| BA|      FTTH|   20419|
| MG|     Wi-Fi|   20292|
| SC|  ETHERNET|   19716|
| CE|      FTTH|   18794|
| RJ|      FTTH|   18542|
| RJ|  ETHERNET|   15170|
| GO|      FTTH|   14857|
| MG|     ADSL1|   14706|
| SP|       HFC|   14256|
+---+----------+--------+
only showing top 20 rows



### Acessos_MovelTratad.parquet

In [ ]:
df2_sp = (spark.read.format("parquet")
                .option("header", True)
                .option("delimiter", ",")
                .load("gs://soulcode-bc17-telecom/Tratados/Acessos_MovelTratad.parquet")
)

In [ ]:
df2_sp.show(4)

+----+---+-------+---+--------------+------------------+----------+-----------------+---------------+-------+-------+-----------------+
| Ano|Mes|Empresa| UF|     Municipio|ModalidadeCobranca|Tecnologia|TecnologiaGeracao|     TipoPessoa|Produto|Acessos|__index_level_0__|
+----+---+-------+---+--------------+------------------+----------+-----------------+---------------+-------+-------+-----------------+
|2021| 12|  CLARO| RS|Estância Velha|          Pós-pago|       LTE|               4G|Pessoa Jurídica|    M2M|      4|                0|
|2021| 12|  CLARO| SP|      Botucatu|          Pós-pago|     WCDMA|               3G|Pessoa Jurídica|    M2M|      5|                1|
|2021| 12|  CLARO| RS|    Sobradinho|          Pós-pago|       LTE|               4G|Pessoa Jurídica|    M2M|      1|                2|
|2021| 12|  CLARO| PR|       Maringá|          Pós-pago|     WCDMA|               3G|Pessoa Jurídica|    M2M|     91|                3|
+----+---+-------+---+--------------+-----------

In [ ]:
df2_sp.select('Ano').distinct().show()

+----+
| Ano|
+----+
|2021|
|2022|
+----+



In [ ]:
df2_sp.createOrReplaceTempView("acesMovel")

### Acessos_Telefonia_Movel_Pre_Pos_TotalTratad.parquet
df3_sp

In [ ]:
df3_sp = (spark.read.format("parquet")
                .option("header", True)
                .option("delimiter", ",")
                .load("gs://soulcode-bc17-telecom/Tratados/Acessos_Telefonia_Movel_Pre_Pos_TotalTratad.parquet")
)

In [ ]:
df3_sp.show(1)

+----+---+------------------+---------+-----------------+
| Ano|Mes|ModalidadeCobranca|  Acessos|__index_level_0__|
+----+---+------------------+---------+-----------------+
|2022|  3|          Pós-pago|138629611|                0|
+----+---+------------------+---------+-----------------+
only showing top 1 row



In [ ]:
df3_sp.select('mes').distinct().show()

+---+
|mes|
+---+
|  7|
|  6|
|  9|
|  5|
|  1|
| 10|
|  3|
| 12|
|  8|
| 11|
|  2|
|  4|
+---+



In [ ]:
df3_sp.createOrReplaceGlobalTempView("acesmovel")

### Densidade_Telefonia_Movel.parque
df4_sp

In [ ]:
df4_sp = (spark.read.format("parquet")
                .option("header", True)
                .option("delimiter", ",")
                .load("gs://soulcode-bc17-telecom/Tratados/Densidade_Telefonia_Movel.parquet")
)

In [ ]:
df4_sp.show(1)

+----+---+------+---------+--------+---------------+---------------+-----------------+
| Ano|Mes|    UF|Municipio|Cod_IBGE|      Densidade|NivelGeografico|__index_level_0__|
+----+---+------+---------+--------+---------------+---------------+-----------------+
|2022|  3|Brasil|   Brasil|     0.0|121,96723524202|         Brasil|                0|
+----+---+------+---------+--------+---------------+---------------+-----------------+
only showing top 1 row



In [ ]:
df4_sp.select('Mes').distinct().show()

+---+
|Mes|
+---+
|  7|
|  6|
|  9|
|  5|
|  1|
| 10|
|  3|
| 12|
|  8|
| 11|
|  2|
|  4|
+---+



In [ ]:
df4_sp.createOrReplaceTempView("densTel")

### Meu_Municipio_AcessosTratad.parquet
df5_sp

In [ ]:
df5_sp = (spark.read.format("parquet")
                .option("header", True)
                .option("delimiter", ",")
                .load("gs://soulcode-bc17-telecom/Tratados/Meu_Municipio_AcessosTratad.parquet")
)

In [ ]:
df5_sp.show(1)

+----+---+-------+---------------+---------------+--------+-------------+---+------+
| Ano|Mes|Acessos|        Servico|      Densidade|Cod_IBGE|    Municipio| UF|Regiao|
+----+---+-------+---------------+---------------+--------+-------------+---+------+
|2021|  6|   5597|Telefonia Móvel|73,114931636659| 4118709|Paulo Frontin| PR|   Sul|
+----+---+-------+---------------+---------------+--------+-------------+---+------+
only showing top 1 row



In [ ]:
df5_sp.select('Mes').distinct().show()

+---+
|Mes|
+---+
|  6|
| 12|
+---+



In [ ]:
df5_sp.createOrReplaceTempView("MeuAcess")

### Velocidade_Contratada_Tratad.parquet
df6_sp

In [ ]:
df6_sp = (spark.read.format("parquet")
                .option("header", True)
                .option("delimiter", ",")
                .load("gs://soulcode-bc17-telecom/Tratados/Velocidade_Contratada_Tratad.parquet")
)

In [ ]:
df6_sp.show(1)

+------+----+-----------------+------------------+--------------------------+---+----------+---------+-------+--------+-----------------+
|   Ano| Mes|     razao_social|              cnpj|velocidade_contratada_mbps| uf| municipio| Cod_IBGE|acessos|    tipo|__index_level_0__|
+------+----+-----------------+------------------+--------------------------+---+----------+---------+-------+--------+-----------------+
|2020.0|12.0|algar telecom s/a|71.208.516/0001-74|                        40| MG|UBERLANDIA|3170206.0|16660.0|INTERNET|                0|
+------+----+-----------------+------------------+--------------------------+---+----------+---------+-------+--------+-----------------+
only showing top 1 row



In [ ]:
df6_sp.select('Ano').distinct().show()                                                                                                                                    

+------+
|   Ano|
+------+
|2020.0|
|2019.0|
|2017.0|
|2018.0|
+------+

